iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/region/PartiteGraphPlacementDestNodeSelector.java

```diff
@@ -0,0 +1,166 @@
+      List<TRegionReplicaSet> databaseAllocatedRegionGroups,
+      int replicationFactor,
+      TConsensusGroupId consensusGroupId,
+      TRegionReplicaSet remainReplicaSet) {
+    prepare(replicationFactor, availableDataNodeMap, allocatedRegionGroups, remainReplicaSet);
+
+    searchTargetDataNode(freeDiskSpaceMap);
+
+    return availableDataNodeMap.get(fakeToRealIdMap.get(targetDataNode));
+  }
+
+  private void prepare(
+      int replicationFactor,
+      Map<Integer, TDataNodeConfiguration> availableDataNodeMap,
+      List<TRegionReplicaSet> allocatedRegionGroups,
+      TRegionReplicaSet remainReplicaSet) {
+    this.replicationFactor = replicationFactor;
+    this.fakeToRealIdMap = new TreeMap<>();
+    this.realToFakeIdMap = new TreeMap<>();
+    this.dataNodeNum = availableDataNodeMap.size();
+    List<Integer> dataNodeIdList =
+        availableDataNodeMap.values().stream()
+            .map(c -> c.getLocation().getDataNodeId())
+            .collect(Collectors.toList());
+    for (int i = 0; i < dataNodeNum; i++) {
+      fakeToRealIdMap.put(i, dataNodeIdList.get(i));
+      realToFakeIdMap.put(dataNodeIdList.get(i), i);
+    }
+
+    remainDataNodesFakeId =
+        remainReplicaSet.dataNodeLocations.stream()
+            .map(TDataNodeLocation::getDataNodeId)
+            .map(realToFakeIdMap::get)
+            .collect(Collectors.toSet());
+
+    // Compute regionCounter and combinationCounter
+    this.regionCounter = new int[dataNodeNum];
+    Arrays.fill(regionCounter, 0);
+    this.combinationCounter = new int[dataNodeNum][dataNodeNum];
+    for (int i = 0; i < dataNodeNum; i++) {
+      Arrays.fill(combinationCounter[i], 0);
+    }
+    for (TRegionReplicaSet regionReplicaSet : allocatedRegionGroups) {
+      List<TDataNodeLocation> dataNodeLocations = regionReplicaSet.getDataNodeLocations();
+      for (int i = 0; i < dataNodeLocations.size(); i++) {
+        int fakeIId = realToFakeIdMap.get(dataNodeLocations.get(i).getDataNodeId());
+        regionCounter[fakeIId]++;
+        for (int j = i + 1; j < dataNodeLocations.size(); j++) {
+          int fakeJId = realToFakeIdMap.get(dataNodeLocations.get(j).getDataNodeId());
+          combinationCounter[fakeIId][fakeJId] = 1;
+          combinationCounter[fakeJId][fakeIId] = 1;
+        }
+      }
+    }
+
+    this.bestValue = new Pair<>(Integer.MAX_VALUE, Integer.MAX_VALUE);
+  }
+
+  private void searchTargetDataNode(Map<Integer, Double> freeDiskSpaceMap) {
+    List<DataNodeEntry> entryList = new ArrayList<>();
+
+    for (int i = 0; i < dataNodeNum; i++) {
+      if (remainDataNodesFakeId.contains(i)) {
+        continue;
+      }
+      entryList.add(
+          new DataNodeEntry(i, regionCounter[i], freeDiskSpaceMap.get(fakeToRealIdMap.get(i))));
+    }
+    Collections.sort(entryList);
+
+    int[] alternativeReplica = new int[replicationFactor];
+
+    int pos = 0;
+    for (Integer integer : remainDataNodesFakeId) {
+      alternativeReplica[pos] = integer;
```

Comment:
It is desirable to add some protection mechanisms to improve robustness in future scenarios where the number of replicas can change dynamically

Reply:
If you want to support future scenarios where the number of replicas can dynamically change, it is best to systematically refactor the entire system.

Reply:
This is not a future work, currently we support extend region so that the number of replicas of regiongroup is not consistent with the configuration parameter. This feeling can be guarded a little more, because otherwise we could end up with an index out of bound error even during region migration if the number of replicas is not consistent

Reply:
added.